sirepo/sim_db_file.py

```diff
@@ -1,48 +1,378 @@
+
+    def _post(self, lib_sid_uri, basename=None, args=None, sim_type=None):
+        res = pkjson.load_any(
+            self._request(
+                "POST",
+                lib_sid_uri,
+                basename,
+                json=PKDict(
+                    method=pkinspect.caller_func_name(),
+                    args=PKDict() if args is None else args,
+                ),
+                sim_type=sim_type,
+            ).content,
+        )
+        if res.get("state") != "ok":
+            raise AssertionError(
+                "expected state=ok reply={} uri={} basename={}", res, basename
+            )
+        return res.get("result")
+
+    def _request(
+        self, method, lib_sid_uri, basename, data=None, json=None, sim_type=None
+    ):
+        u = self.uri(lib_sid_uri, basename, sim_type)
+        r = sirepo.agent_supervisor_api.request(
+            method,
+            _cfg.server_uri + u,
+            _cfg.server_token,
+            data=data,
+            json=json,
+        )
+        if r.status_code == 404:
+            raise FileNotFoundError(f"sim_db_file={u} not found")
+        r.raise_for_status()
+        return r
+
+
+class SimDbServer(sirepo.agent_supervisor_api.ReqBase):
     _TOKEN_TO_UID = PKDict()
     _UID_TO_TOKEN = PKDict()
 
-    def delete(self, path):
-        for f in pkio.sorted_glob(f"{self.__authenticated_path()}*"):
+    async def get(self, unused_arg):
+        if not self.__authenticate_and_path():
+            return
+        if self.__path.exists():
+            self.write(pkio.read_binary(self.__path))
+        else:
+            self.send_error(404)
+
+    async def post(self, unused_arg):
+        def _result(value):
+            if value is None:
+                value = PKDict()
+            elif not isinstance(value, PKDict):
+                value = PKDict(result=value)
+            return value.pksetdefault(state="ok")
+
+        try:
+            if not self.__authenticate_and_path():
+                return
+            r = pkjson.load_any(self.request.body)
+            # note that args may be empty (but must be PKDict), since uri has path
+            if not isinstance(a := r.get("args"), PKDict):
+                raise AssertionError(f"invalid post path={self.__path} args={a}")
+            if not (m := r.get("method")):
+                raise AssertionError(f"missing method path={self.__path} args={a}")
+            self.write(_result(await getattr(self, "_sr_post_" + m)(a)))
+        except Exception as e:
+            if pkio.exception_is_not_found(e):
+                self.send_error(404)
+                return
+            pkdlog(
+                "uri={} body={} exception={} stack={}",
+                self.request.path,
+                self.request.body,
+                e,
+                pkdexc(),
+            )
+            try:
+                self.write({state: "error"})
+            except Exception:
```

Comment:
Why wouldn't we want to blow up?

Reply:
This is go-style. The None cascades to a place where something can do something about it. The error context is printed correctly (pkdlog logs the line), but the server returns 403.

Reply:
> I didn't understand your response. I probably just don't understand this code...
>> The None cascades to a place where something can do something about it
> If there is a None being used in a meaningful way here please be explicit about it. I can't tell from looking at the code that there is a meaningful use of None.

I was thinking about different code when I wrote this. The reason for the pass here is that there's something wrong with tornado. Instead I logged the error 11793fdc3920d87ea957f47a361c3502f8e1111a. I don't like just dropping into tornado and hoping it handles the problem. I think we should understand all errors, and translate them to replies ourselves.

This particular code will be very hard to test.

>> The error context is printed correctly (pkdlog logs the line)
> If there is an error with write we don't log (as far as I can tell). That's what I'm curious about. Why would we want to swallow that error?

Apologies for the confusion.